systems/framework/lcm_driven_loop.h

```diff
@@ -4,90 +4,207 @@
+                const std::string& active_channel)
+      : LcmDrivenLoop(drake_lcm, std::move(diagram), lcm_parser) {
+    DRAKE_DEMAND(!input_channels.empty());
+    is_multiple_inputs_ = true;
+
+    // Create subscriber for the switch
+    switch_sub_ = std::make_unique<drake::lcm::Subscriber<SwitchMessageType>>(
+        drake_lcm_, switch_channel);
+
+    // Create subscribers for inputs
+    for (const auto& name : input_channels) {
+      std::cout << "Constructing subscriber for " << name << std::endl;
+      name_to_input_sub_map_.insert(std::make_pair(
+          name, drake::lcm::Subscriber<InputMessageType>(drake_lcm_, name)));
+    }
+
+    // Default initial active channel
+    active_channel_ = active_channel;
+  };
 
   // Start simulating the diagram
-  void Simulate(double end_time = std::numeric_limits<double>::infinity());
-
-  // Give a name to the diagram
-  void SetName(const std::string & diagram_name);
+  void Simulate(double end_time = std::numeric_limits<double>::infinity()) {
+    // Get mutable contexts
+    auto& diagram_context = simulator_->get_mutable_context();
+    auto& lcm_parser_context = diagram_ptr_->GetMutableSubsystemContext(
+        *lcm_parser_, &diagram_context);
+
+    // Wait for the first message.
+    drake::log()->info("Waiting for first lcm input message");
+    LcmHandleSubscriptionsUntil(drake_lcm_, [&]() {
+      return name_to_input_sub_map_.at(active_channel_).count() > 0;
+    });
+
+    // Initialize the context based on the first message.
+    const double t0 =
+        name_to_input_sub_map_.at(active_channel_).message().utime * 1e-6;
+    diagram_context.SetTime(t0);
+    auto& input_value = lcm_parser_->get_input_port(0).FixValue(
+        &lcm_parser_context,
+        name_to_input_sub_map_.at(active_channel_).message());
+
+    // "Simulator" time
+    double time = 0;  // initialize the current time with 0
+    double message_time;
+
+    drake::log()->info(diagram_name_ + " started");
+    // Run the simulation until end_time
+    while (time < end_time) {
+      // Update the name of the active channel if there are multiple inputs
+      if (is_multiple_inputs_) {
+        if (switch_sub_->count() > 0) {
+          // Check if the channel name is a key of the map. If it is, we update
+          // the active channel name and clear switch_sub_'s message. If it is
+          // not we do not update the active channel name.
+          if (name_to_input_sub_map_.count(switch_sub_->message().channel) ==
+              1) {
+            active_channel_ = switch_sub_->message().channel;
+          } else {
+            std::cout << switch_sub_->message().channel << " doesn't exist\n";
+          }
+          switch_sub_->clear();
+        }
+      }
+
+      // Wait for an InputMessageType message.
+      name_to_input_sub_map_.at(active_channel_).clear();
+      LcmHandleSubscriptionsUntil(drake_lcm_, [&]() {
+        return (name_to_input_sub_map_.at(active_channel_).count() > 0);
```

Comment:
Shouldn't this exit on either the `active_channel` subscriber or the switching subscriber, with a different result depending on which is triggered?

Reply:
I thought about that, but it makes the logic more complicated (e.g. we need a variable to remember the previous message), and do you think it's necessary?

Reply:
I think it's a big advantage, as it will allow us to potentially recover from controllers that crash.

Why does it need to remember the previous message?

Reply:
Ok, the feature is added. I was wrong that we need to remember the previous message. In the implementation, I just don't update the message value when there is no new input message.